feat: Add "unrecognized key in configuration file" warning

ajita-asthana opened the pull request:
This PR adds a warning when encountered with `unrecognized_value` for a tile source. 

Resolves #1838

This change doesn't include the integration tests.

ajita-asthana committed:
add a warning for unrecognized values for a tile source

ajita-asthana committed:
Merge branch 'maplibre:main' into unrecognized_warning

ajita-asthana committed:
add map related API

ajita-asthana committed:
add missing semicolon

CommanderStorm commented on martin/src/file_config.rs:
```diff
@@ -171,6 +171,9 @@ impl<T: ConfigExtras> FileConfigEnum<T> {
         }
+        if res.is_empty() {
+            warn!("There exists an unrecognized value");
+        }
```
Two things:
- this warning is not really actionable. We need to tell people at minimum which key is unregognised
- Would it not be better to add a warning one level up, i.e. where this is called?

ajita-asthana replied:
I see. I thought if `res` is empty then the value would be triggered. 
side:
I was looking around for adding the test case line in each block of yaml file but it didn't make sense to me.

ajita-asthana committed:
Merge branch 'maplibre:main' into unrecognized_warning

ajita-asthana committed:
verbose warning

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm commented on martin/src/file_config.rs:
```diff
@@ -168,8 +168,18 @@ impl<T: ConfigExtras> FileConfigEnum<T> {
+                    prefix,
+                    cfg.get_unrecognized()
+                );
+            }
```
Much better.

I think below code could be a bit more user friendly.
This way, we are telling people less abstract which value is unexpected.

This also fixes the duplicate reporting bug caused by checking `res` instead of `unrecognized`.

```suggestion
            let unrecognized = cfg.get_unrecognized();
            copy_unrecognized_config(&mut res, prefix, unrecognized);

            for (key, value) in unrecognized.into_iter() {
                warn!("Ignoring unrecognized configuration entry '{prefix}.{key}: {value}'. Please check your configuration file for typos.");
            }
```

ajita-asthana replied:
Thanks @CommanderStorm .

ajita-asthana committed:
Update martin/src/file_config.rs

Co-authored-by: Frank Elsinga <frank@elsinga.de>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ajita-asthana committed:
add a test case

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyurik committed:
Merge branch 'main' into unrecognized_warning

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm commented on martin/src/file_config.rs:
```diff
@@ -168,8 +168,16 @@ impl<T: ConfigExtras> FileConfigEnum<T> {
+            for (key, value) in unrecognized.into_iter() {
+                warn!(
+                    "Ignoring unrecognized configuration entry '{prefix}.{key}: {value}'. Please check your configuration file for typos."
+                );
```
There is a clippy warning about `value` not having `Display` implemented.

Also, on second thought, we are not really objecting to `value`, rather we are objecting on the `key` not being something we recognise.

```suggestion
            for key in unrecognized.keys() {
                warn!(
                    "Ignoring unrecognized configuration key '{prefix}.{key}'. Please check your configuration file for typos."
                );
```

CommanderStorm commented on tests/config.yaml:
```diff
@@ -169,29 +169,35 @@ postgres:
 
 
 pmtiles:
+  this-should: be-a-warning
```
That is a good start 👍🏻

- there are more parts where we should add this above
- we now need to assert that the warning has been raised in `test.sh`

CommanderStorm committed:
chore(ci): don't accidentally automerge due to skipped checks (#1946)

This fixes the stupid problem that our CI allows skipping it in some
circumstances

> [!NOTE]
> A job that is skipped will report its status as "Success". It will not
prevent a pull request from merging, even if it is a required check.


https://docs.github.com/en/actions/how-tos/writing-workflows/choosing-when-your-workflow-runs/using-conditions-to-control-job-execution

It turns out the reason a lot of people do this is more underlying

Resolves #1811

CommanderStorm committed:
chore(core): make sure that dependencys are clearly attributed (#1947)

During the discovery-PR of what the next best step in pulling into
`martin-core` is, I discovered, that our dependency-feature gating could
be tighter.

The primary motivation behind this PR is that if what feature has which
dependency is well-defined, migration is much simpler.

This PR was tested via 

```bash
cargo hack check --each-feature --keep-going --locked
```

CommanderStorm committed:
chore(core): move `TileRect` to `martin-tile-utils` (#1945)

This is part of the `martin-core` refactoring.

This utility should really be a tile-utility.

CommanderStorm committed:
chore: fix todo by migrating `OnceLock` to `LazyLock` (#1948)

During other work I noticed this TODO, was simple enough to fix

CommanderStorm committed:
chore(core): make sure that there is only one `TileData` (#1951)

this is not actively a problem, just something that I noticed during the
refactoring excercise towards `martin-core` which also spawned
https://github.com/maplibre/martin/pull/1949

CommanderStorm committed:
chore(ci): simplify the `mdbook-alerts` installation (#1952)

This PR changes us to have a simpler `mdbook-alerts` installation.

Requires the following PR to pass CI:
- https://github.com/taiki-e/install-action/pull/1060

CommanderStorm committed:
chore(core): unify where we handle config (#1950)

Split out the config into dedicated files

---------

Co-authored-by: Yuri Astrakhan <yuriastrakhan@gmail.com>

CommanderStorm committed:
feat(ui): Add an analytics section via `/_/metrics` (#1942)

This draft breaks out just the analytics section of the frontend

It is a bit larger, but most of the changes are tests, so likey fine.
The big part is tests.

The commits on this one are a bit of a mess, but I am scared to squash
at this point. The rebase would likely result in at least one bug.

# Screenshot

<img width="1444" height="762" alt="image"
src="https://github.com/user-attachments/assets/e6fbd312-b00c-4ba1-927c-b482c958b7dc"
/>

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Yuri Astrakhan <yuriastrakhan@gmail.com>

CommanderStorm committed:
fix(ui): fix infinte loop on the dasboad metrics section (#1953)

during one of the merges of #1897, I seem to have mismerged.

This PR restores us to the state that I wanted #1942 to be

<img width="1528" height="650" alt="image"
src="https://github.com/user-attachments/assets/a2a5a136-03ee-4a94-9a23-e538767e9fda"
/>

nyurik committed:
chore: sort justfile (#1960)

CommanderStorm committed:
feat(ui): Add the font catalog to the dashboard (#1954)

This breaks the font catalog out of #1897 

Yes, this seems also quite heavy in terms of lines chaneged, but this is
the smallest version of this feature sadly.
I hope that from a reviewer load, this is still acceptable though.

It WIP since I still need to do my own review

<img width="1470" height="953" alt="image"
src="https://github.com/user-attachments/assets/3bfed819-cbba-4f18-8435-21235ca164b4"
/>

CommanderStorm committed:
feat(ui): Add sprite catalog without dialogs to the dashboard (#1955)

> [!IMPORTANT]
> This PR bases on https://github.com/maplibre/martin/pull/1954 Said PR
should be reviewed first.

This PR implements the sprite catalog.

To remain managable, I have split this one in two.

<img width="1514" height="657" alt="image"
src="https://github.com/user-attachments/assets/c6c507a8-0982-4c70-9527-4ad07cbe6831"
/>

CommanderStorm committed:
feat(ui): Sprite catalog with dialogs for the dashboard (#1956)

This PR implements the full dialogs for the sprite catalog.

<img width="1822" height="914" alt="image"
src="https://github.com/user-attachments/assets/dbb62987-6f53-4cae-84b6-bb569f3b9c10"
/>

CommanderStorm committed:
Apply code suggestion

CommanderStorm committed:
added the warning messages to where they should be

CommanderStorm committed:
implement warning for unspecified items

CommanderStorm committed:
Merge branch 'main' into unrecognized_warning

CommanderStorm committed:
fix a clippy warning

CommanderStorm committed:
improve how `get_unrecognized_keys` is handled

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm committed:
fix import related typos

CommanderStorm committed:
Merge branch 'main' into unrecognized_warning

CommanderStorm committed:
Merge branch 'main' into unrecognized_warning

CommanderStorm committed:
Merge branch 'main' into document-SourceConfigExtras

CommanderStorm committed:
Merge branch 'main' into document-SourceConfigExtras

CommanderStorm committed:
fix unused import

CommanderStorm committed:
update blessed output

CommanderStorm committed:
make sure that unrecognised values are never serialized

CommanderStorm committed:
Merge branch 'main' into unrecognized_warning

nyurik committed:
Merge branch 'main' into document-SourceConfigExtras

CommanderStorm committed:
Merge branch 'main' into document-SourceConfigExtras

nyurik committed:
Merge branch 'main' into document-SourceConfigExtras

CommanderStorm committed:
Merge branch 'main' into document-SourceConfigExtras

CommanderStorm committed:
fix merge issue

CommanderStorm committed:
Merge branch 'document-SourceConfigExtras' into unrecognized_warning

CommanderStorm committed:
Merge branch 'main' into unrecognized_warning

CommanderStorm committed:
fix a few merge issues

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
